src/mem/localalloc.h

```diff
@@ -449,6 +449,133 @@ namespace snmalloc
+       * edge case around monotonicity.
+       */
+      check_client(
+        __builtin_cheri_length_get(p) > 0,
+        "Zero-length capability in deallocation");
+
+      /*
+       * At present we check for the pointer also being the start of an
+       * allocation closer to dealloc; for small objects, that happens in
+       * dealloc_local_object_fast, either below or *on the far end of message
+       * receipt*.  For large objects, it happens below by directly rounding to
+       * power of two rather than using the is_start_of_object helper.
+       * (XXX This does mean that we might end up threading our remote queue
+       * state somewhere slightly unexpected rather than at the head of an
+       * object.  That is perhaps fine for now?)
+       */
+
+      /*
+       * TODO
+       *
+       * We could enforce other policies here, including that the length exactly
+       * match the sizeclass.  At present, we bound caps we give for allocations
+       * to the underlying sizeclass, so even malloc(0) will have a non-zero
+       * length.  Monotonicity would then imply that the pointer must be the
+       * head of an object (modulo, perhaps, temporal aliasing if we somehow
+       * introduced phase shifts in heap layout like some allocators do).
+       *
+       * If we switched to bounding with upwards-rounded representable bounds
+       * (c.f., CRRL) rather than underlying object size, then we should,
+       * instead, in general require plausibility of p_raw by checking that its
+       * length is nonzero and the snmalloc size class associated with its
+       * length is the one for the slab in question... except for the added
+       * challenge of malloc(0).  Since 0 rounds up to 0, we might end up
+       * constructing zero-length caps to hand out, which we would then reject
+       * upon receipt.  Instead, as part of introducing CRRL bounds, we should
+       * introduce a sizeclass for slabs holding zero-size objects.  All told,
+       * we would want to check that
+       *
+       *   size_to_sizeclass(length) == entry.get_sizeclass()
+       *
+       * I believe a relaxed CRRL test of
+       *
+       *   length > 0 || (length == sizeclass_to_size(entry.get_sizeclass()))
+       *
+       * would also suffice and may be slightly less expensive than the test
+       * above, at the cost of not catching as many misbehaving clients.
+       *
+       * In either case, having bounded by CRRL bounds, we would need to be
+       * *reconstructing* the capabilities headed to our free lists to be given
+       * out to clients again; there are many more CRRL classes than snmalloc
+       * sizeclasses (this is the same reason that we can always get away with
+       * CSetBoundsExact in capptr_bound).  Switching to CRRL bounds, if that's
+       * ever a thing we want to do, will be easier after we've done the
+       * plumbing for CHERI+MTE.
+       */
```

Comment:
In the CheriBSD jemalloc adaptation we do return zero-sized caps for `malloc(0)`. We also use CRRL bounds. This does complicate potential checks of capabilities passed to free as one can only check if the length is plausible unless a set of valid lengths is stored somewhere. This should probably be revisited.